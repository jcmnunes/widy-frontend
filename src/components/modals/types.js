/* eslint-disable import/prefer-default-export */
export const ADD_TASK = 'modal/ADD_TASK';
