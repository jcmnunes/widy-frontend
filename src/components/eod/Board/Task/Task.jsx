import React from 'react';
import PropTypes from 'prop-types';
import styled, { keyframes } from 'styled-components';
import { CopyToClipboard } from 'react-copy-to-clipboard';
import noop from 'lodash.noop';
import Timer from '../../Timer';
import TaskMenu from '../TaskMenu';
import { Checkbox } from '../../../UI';
import { IconDuplicate } from '../../../../icons/Icons';
import toast from '../../../../helpers/toast';

const getColors = props => {
  const colors = {
    border: props.theme.neutral200,
    background: 'white',
  };

  if (props.isSelected) {
    colors.border = props.theme.yellow500;
    colors.background = props.theme.yellow050;
  }

  if (props.isInBreak) {
    colors.border = props.theme.blue200;
  }

  if (props.isCompleted) {
    colors.border = props.theme.neutral100;
  }

  if (props.isDragging) {
    colors.background = props.theme.blue050;
    colors.border = props.theme.blue050;
  }

  return colors;
};

const pulseAnimationIsActive = props => keyframes`
  from {
    border-color: ${props.theme.yellow100};
    box-shadow: 0 0 0 4px ${props.theme.yellow050};
  }

  to {
    border-color: ${props.theme.yellow700};
    box-shadow: 0 0 0 4px ${props.theme.yellow200};
  }
`;

const pulseAnimationInBreak = props => keyframes`
  from {
    border-color: ${props.theme.blue050};
    box-shadow: 0 0 0 4px ${props.theme.blue050};
  }

  to {
    border-color: ${props.theme.blue400};
    box-shadow: 0 0 0 4px ${props.theme.blue100};
  }
`;

const getAnimation = props => {
  if (props.isInBreak) return pulseAnimationInBreak(props);
  if (props.isActive) return pulseAnimationIsActive(props);
  return null;
};

const StyledIconDuplicate = styled(IconDuplicate)`
  display: none;
  cursor: pointer;
`;

const StyledTask = styled.div`
  display: flex;
  align-items: center;
  flex-direction: row;
  border: ${props => `1px solid ${getColors(props).border}`};
  border-radius: 4px;
  background-color: ${props => getColors(props).background};
  padding: 8px;
  font-size: 16px;
  margin: 4px 0;
  color: ${props => (props.isCompleted ? props.theme.neutral300 : props.theme.neutral700)};
  text-decoration: ${props => (props.isCompleted ? 'line-through' : 'none')};
  cursor: pointer;
  animation: ${props => getAnimation(props)};
  animation-duration: 1s;
  animation-iteration-count: infinite;
  animation-direction: alternate;

  &:hover {
    background-color: ${props =>
      props.isSelected ? props.theme.yellow075 : props.theme.neutral025};

    ${StyledIconDuplicate} {
      display: inline-block;
    }
  }
`;

const TaskName = styled.span`
  flex: 1;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  width: 0;
  margin-right: 16px;
`;

const Controls = styled.div`
  display: flex;
  flex-direction: row;

  > * {
    margin-left: 4px;
  }
`;

const Control = styled.div`
  width: 25px;
  height: 25px;
  border-radius: 4px;
  background-color: rgba(0, 0, 0, 0);
  transition: background-color 0.2s ease;

  &:hover {
    background-color: ${props => props.theme.neutral100};
  }
`;

const Task = ({
  taskRef,
  taskId,
  sectionId,
  isSelected,
  isActive,
  isInBreak,
  isCompleted,
  renderControls,
  onClick,
  onDoubleClick,
  onCheckChange,
  onCheckClick,
  storeSelectedSectionId,
  storeSelectedTaskId,
  openModal,
  isDragging,
  children,
  ...other
}) => (
  <StyledTask
    ref={taskRef}
    isSelected={isSelected}
    isActive={isActive}
    isInBreak={isInBreak}
    isCompleted={isCompleted}
    onClick={onClick}
    storeSelectedSectionId={storeSelectedSectionId}
    isDragging={isDragging}
    {...other}
  >
    <Checkbox checked={isCompleted} onChange={onCheckChange} onClick={onCheckClick} />
    <TaskName onDoubleClick={onDoubleClick}>{children}</TaskName>
    <CopyToClipboard
      text={children}
      onCopy={() =>
        toast.success({
          title: 'Success',
          message: 'Task title copied',
        })
      }
    >
      <StyledIconDuplicate />
    </CopyToClipboard>
    {renderControls && !isCompleted && (
      <Controls>
        <Timer taskId={taskId} sectionId={sectionId} />
        <Control>
          <TaskMenu
            taskId={taskId}
            sectionId={sectionId}
            storeSelectedSectionId={storeSelectedSectionId}
            storeSelectedTaskId={storeSelectedTaskId}
            openModal={openModal}
          />
        </Control>
      </Controls>
    )}
  </StyledTask>
);

Task.defaultProps = {
  taskRef: null,
  renderControls: true,
  onClick: noop,
  onDoubleClick: noop,
  onCheckChange: noop,
  onCheckClick: noop,
  storeSelectedTaskId: noop,
  storeSelectedSectionId: noop,
  openModal: noop,
};

Task.propTypes = {
  taskRef: PropTypes.func,
  renderControls: PropTypes.bool,
  isCompleted: PropTypes.bool.isRequired,
  taskId: PropTypes.string.isRequired,
  sectionId: PropTypes.string.isRequired,
  isSelected: PropTypes.bool.isRequired,
  isActive: PropTypes.bool.isRequired,
  isInBreak: PropTypes.bool.isRequired,
  children: PropTypes.string.isRequired,
  onClick: PropTypes.func,
  onDoubleClick: PropTypes.func,
  onCheckChange: PropTypes.func,
  onCheckClick: PropTypes.func,
  storeSelectedTaskId: PropTypes.func,
  storeSelectedSectionId: PropTypes.func,
  openModal: PropTypes.func,
  isDragging: PropTypes.bool.isRequired,
};

export default Task;
