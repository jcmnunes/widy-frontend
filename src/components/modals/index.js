export { default as AddTask } from './AddTask';
export { default as LaunchTask } from './LaunchTask';
export { default as RenameTask } from './RenameTask';
