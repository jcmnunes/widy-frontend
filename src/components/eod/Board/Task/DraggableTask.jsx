import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Draggable } from 'react-beautiful-dnd';
import Task from './Task';
import { RENAME_TASK } from '../../../modals/types';

class DraggableTask extends Component {
  handleTaskClick = () => {
    const { sectionId, taskId } = this.props;
    this.props.storeSelectedSectionId(sectionId);
    this.props.storeSelectedTaskId(taskId);
    this.props.openSidebar();
  };

  handleTaskDoubleClick = () => {
    const { sectionId, taskId } = this.props;
    this.props.storeSelectedSectionId(sectionId);
    this.props.storeSelectedTaskId(taskId);
    this.props.openModal(RENAME_TASK);
  };

  handleTaskCheckboxChange = () => {
    const { activeTask, sectionId, taskId, isCompleted } = this.props;
    this.props.storeSelectedSectionId(sectionId);
    this.props.storeSelectedTaskId(taskId);
    if (activeTask.taskId === taskId) {
      this.props.stopTask(taskId, sectionId);
    }
    this.props.updateTask(taskId, { completed: !isCompleted, start: null });
  };

  handleTaskCheckboxClick = e => {
    e.stopPropagation();
  };

  render() {
    const {
      taskId,
      sectionId,
      selectedTaskId,
      activeTask,
      index,
      isCompleted,
      taskTitle,
    } = this.props;
    return (
      <Draggable draggableId={taskId} index={index}>
        {(provided, snapshot) => (
          <Task
            taskRef={provided.innerRef}
            {...provided.draggableProps}
            {...provided.dragHandleProps}
            taskId={taskId}
            sectionId={sectionId}
            isSelected={taskId === selectedTaskId}
            isCompleted={isCompleted}
            isActive={activeTask.taskId === taskId && !activeTask.inBreak}
            isInBreak={activeTask.taskId === taskId && activeTask.inBreak}
            onClick={this.handleTaskClick}
            onDoubleClick={this.handleTaskDoubleClick}
            onCheckChange={this.handleTaskCheckboxChange}
            onCheckClick={this.handleTaskCheckboxClick}
            storeSelectedSectionId={this.props.storeSelectedSectionId}
            storeSelectedTaskId={this.props.storeSelectedTaskId}
            openModal={this.props.openModal}
            isDragging={snapshot.isDragging}
          >
            {taskTitle}
          </Task>
        )}
      </Draggable>
    );
  }
}

DraggableTask.propTypes = {
  index: PropTypes.number.isRequired,
  sectionId: PropTypes.string.isRequired,
  taskId: PropTypes.string.isRequired,
  selectedTaskId: PropTypes.string.isRequired,
  activeTask: PropTypes.shape({
    taskId: PropTypes.string.isRequired,
    inBreak: PropTypes.bool.isRequired,
  }).isRequired,
  isCompleted: PropTypes.bool.isRequired,
  taskTitle: PropTypes.string.isRequired,
  storeSelectedTaskId: PropTypes.func.isRequired,
  storeSelectedSectionId: PropTypes.func.isRequired,
  openSidebar: PropTypes.func.isRequired,
  openModal: PropTypes.func.isRequired,
  updateTask: PropTypes.func.isRequired,
  stopTask: PropTypes.func.isRequired,
};

export default DraggableTask;
