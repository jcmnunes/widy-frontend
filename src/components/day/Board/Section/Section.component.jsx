import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Droppable } from 'react-beautiful-dnd';
import Task from '../Task';
import PlanTask from '../PlanTask/PlanTask.container';
import { Heading2 } from '../../../UI/Typography';
import { IllustrationPlan } from '../../../../icons/Illustrations';
import { IconAdd, SectionWithNoTasks, StyledLink, StyledSection, Tasks } from './Section.styles';

class SectionComponent extends Component {
  renderSection = provided => {
    const { section } = this.props;
    return (
      <Tasks ref={provided.innerRef} {...provided.droppableProps}>
        {section.tasks.map((taskId, index) => {
          if (section.isPlan) {
            return <PlanTask key={taskId} taskId={taskId} index={index} sectionId={section.id} />;
          }
          return <Task key={taskId} taskId={taskId} index={index} sectionId={section.id} />;
        })}
        {provided.placeholder}
      </Tasks>
    );
  };

  renderEmptySection = (provided, snapshot) => {
    const { section, noTasks, dayId, sectionId, openCreateTaskModal } = this.props;
    return (
      <SectionWithNoTasks
        ref={provided.innerRef}
        {...provided.droppableProps}
        isDraggingOver={snapshot.isDraggingOver}
        isPlan={section.isPlan}
        onClick={openCreateTaskModal(dayId, sectionId)}
        noTasks={noTasks}
      >
        {section.isPlan && (
          <>
            {noTasks && <IllustrationPlan />}
            {noTasks ? (
              <span>Start by adding tasks here to plan your day.</span>
            ) : (
              <span>No tasks in Plan.</span>
            )}
          </>
        )}
        {!section.isPlan &&
          (snapshot.isDraggingOver ? (
            <span>Add task to section &quot;{section.title}&quot;</span>
          ) : (
            <span>No tasks in section &quot;{section.title}&quot;</span>
          ))}
      </SectionWithNoTasks>
    );
  };

  render() {
    const { section, dayId, sectionId, openCreateTaskModal } = this.props;
    return (
      <StyledSection>
        <Heading2>{section.title}</Heading2>
        <Droppable droppableId={section.id}>
          {(provided, snapshot) =>
            section.tasks.length > 0
              ? this.renderSection(provided)
              : this.renderEmptySection(provided, snapshot)
          }
        </Droppable>
        <StyledLink
          intent="secondary"
          onClick={openCreateTaskModal(dayId, sectionId)}
          data-test="add-task-button"
        >
          <IconAdd /> {section.isPlan ? 'Add to Plan' : 'Add task'}
        </StyledLink>
      </StyledSection>
    );
  }
}

SectionComponent.propTypes = {
  dayId: PropTypes.string.isRequired,
  sectionId: PropTypes.string.isRequired,
  noTasks: PropTypes.bool.isRequired,
  section: PropTypes.shape({
    tasks: PropTypes.array,
    title: PropTypes.string,
  }).isRequired,
  openCreateTaskModal: PropTypes.func.isRequired,
};

export default SectionComponent;
