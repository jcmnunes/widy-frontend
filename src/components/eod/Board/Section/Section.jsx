import React, { Component } from 'react';
import styled from 'styled-components';
import Task from '../Task/Task';
import Link from '../../../UI/Link';
import { ADD_TASK } from '../../../modals/types';
import { IconAdd } from '../../../../icons/Icons';
import theme from '../../../../styles/theme';

const StyledSection = styled.div`
  margin: 32px 0;

  h1.title {
    font-size: 16px;
    font-weight: bold;
    margin-bottom: 12px;
  }
`;

const EmptyTasks = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  border: 1px solid ${props => props.theme.neutral100};
  border-radius: 4px;
  height: 96px;
  margin-bottom: 8px;
  background-color: white;
  font-size: 16px;
  color: ${props => props.theme.neutral300};
  font-style: italic;
`;
EmptyTasks.displayName = 'EmptyTasks';

const Tasks = styled.div`
  display: flex;
  flex-direction: column;
`;
Tasks.displayName = 'Tasks';

class Section extends Component {
  openModal = () => {
    const { dayId, sectionId } = this.props;
    this.props.storeCreateTaskData(dayId, sectionId);
    this.props.openModal(ADD_TASK);
  };

  render() {
    const { section, tasks } = this.props;
    return (
      <StyledSection>
        <h1 className="title">{section.title}</h1>
        {section.tasks.length ? (
          <Tasks>
            {section.tasks.map(id => (
              <Task key={id}>{tasks[id].title}</Task>
            ))}
          </Tasks>
        ) : (
          <EmptyTasks>
            No tasks in section "<span>{section.title}</span>"
          </EmptyTasks>
        )}
        <Link intent="secondary" onClick={this.openModal}>
          <IconAdd primaryColor={theme.neutral500} /> Add task
        </Link>
      </StyledSection>
    );
  }
}

export default Section;
