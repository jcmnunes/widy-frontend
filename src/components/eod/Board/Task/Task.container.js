import { connect } from 'react-redux';
import DraggableTask from './DraggableTask';
import { openSidebar } from '../../../../actions/sidebar';
import { storeSelectedSectionId } from '../../../../actions/sections';
import { storeSelectedTaskId } from '../../../../actions/tasks';
import { openModal } from '../../../../actions/modals';

const mapStateToProps = state => ({
  selectedTaskId: state.tasks.selected,
  activeTask: state.activeTask,
});

export default connect(
  mapStateToProps,
  { openSidebar, storeSelectedTaskId, storeSelectedSectionId, openModal },
)(DraggableTask);
