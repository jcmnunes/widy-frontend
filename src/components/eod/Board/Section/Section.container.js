import { connect } from 'react-redux';
import Section from './Section';
import { openModal } from '../../../../actions/modals';
import { storeCreateTaskData } from '../../../../actions/tasks';

const mapStateToProps = (state, props) => ({
  section: state.sections.byId[props.sectionId],
  tasks: state.tasks.byId,
});

export default connect(
  mapStateToProps,
  { openModal, storeCreateTaskData },
)(Section);
