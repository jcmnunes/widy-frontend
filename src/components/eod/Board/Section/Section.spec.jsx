import React from 'react';
import { shallow } from 'enzyme';
import 'jest-styled-components';

import Section from './Section';
import Task from '../Task';
import fixture from './Section.fixture';
import { ADD_TASK } from '../../../modals/types';

const TestRenderer = require('react-test-renderer');

describe('EOD Components', () => {
  describe('Section', () => {
    const props = {
      ...fixture,
      openModal: jest.fn(),
      storeCreateTaskData: jest.fn(),
    };

    xit('should render as expected, without crashing', () => {
      const tree = TestRenderer.create(<Section {...props} />).toJSON();
      expect(tree).toMatchSnapshot();
    });

    xit('should open a model when clicking the add task button', () => {
      const wrapper = shallow(<Section {...props} />);
      wrapper.find('Link').simulate('click');
      expect(props.openModal).toHaveBeenCalledWith(ADD_TASK);
    });

    xit('should store the dayId and sectionId when adding a new task', () => {
      const wrapper = shallow(<Section {...props} />);
      wrapper.find('Link').simulate('click');
      expect(props.storeCreateTaskData).toHaveBeenCalledWith(fixture.dayId, fixture.sectionId);
    });

    xit('should render an "EmptyTasks" component when there are no tasks to render', () => {
      const wrapper = shallow(<Section {...props} />);
      expect(wrapper.find('EmptyTasks')).toHaveLength(1);
    });

    xit('should render a list of tasks when there are tasks to render', () => {
      const newProps = {
        ...props,
        section: {
          ...props.section,
          tasks: ['5c9ee9369460ec33384b67df', '5c9ee9369460ec33384b67hf'],
        },
        tasks: {
          '5c9ee9369460ec33384b67df': {
            title: 'Task 1',
          },
          '5c9ee9369460ec33384b67hf': {
            title: 'Task 2 ',
          },
        },
      };
      const wrapper = shallow(<Section {...newProps} />);
      expect(wrapper.find('Tasks')).toHaveLength(1);
      expect(wrapper.find(Task)).toHaveLength(2);
      expect(wrapper.find('EmptyTasks')).toHaveLength(0);
    });
  });
});
