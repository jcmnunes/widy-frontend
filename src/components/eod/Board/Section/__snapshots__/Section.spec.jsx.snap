// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Section should render as expected, without crashing 1`] = `
.c3 {
  font-size: 16px;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: none;
  cursor: pointer;
  padding-left: 0;
  margin-left: 4px;
}

.c3:hover {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c3:focus {
  outline: none;
  box-shadow: 0 0 0 4px;
}

.c3:disabled {
  cursor: not-allowed;
}

.c1 {
  font-size: 16px;
  font-weight: bold;
  margin-bottom: 12px;
}

.c0 {
  margin: 32px 0;
}

.c2 {
  margin-top: 12px;
}

<div
  className="c0"
>
  <h2
    className="c1"
  >
    Title
  </h2>
  <div
    data-test="Droppable"
  >
    Droppable
  </div>
  <button
    className="c2 c3"
    data-test="add-task-button"
    disabled={false}
    onClick={[Function]}
    size="medium"
    type="button"
  >
    <svg
      height={24}
      viewBox="0 0 24 24"
      width={24}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M17 11a1 1 0 0 1 0 2h-4v4a1 1 0 0 1-2 0v-4H7a1 1 0 0 1 0-2h4V7a1 1 0 0 1 2 0v4h4z"
        fill="#616E7C"
        fillRule="evenodd"
      />
    </svg>
     
    Add task
  </button>
</div>
`;
